rfcs/geometric-primitives.md

```diff
@@ -0,0 +1,419 @@
+  points: Vec<Point>,
+}
+impl Meshable for PolyLine {}
+
+struct Triangle([Point; 3]);
+impl Meshable for Triangle {}
+
+struct Quad([Point; 4]);
+impl Meshable for Quad {}
+
+/// A sphere centered on the origin
+struct Sphere{
+  radius: f32,
+}
+impl Meshable for Sphere {}
+
+/// A cylinder with its origin at the center of the volume
+struct Cylinder {
+  radius: f32,
+  height: f32,
+}
+impl Meshable for Cylinder {}
+
+/// A cone with the origin located at the center of the circular base
+struct Cone{
+  /// Origin of the cone located at the center of its base.
+  origin: Point,
+  /// The extent of the cone is the vector that extends from the origin to the tip of the cone.
+  extent: Vec3,
+  base_radius: f32,
+}
+impl Meshable for Cone {}
+
+struct Torus{
+  major_radius: f32,
+  tube_radius: f32,
+}
+impl Meshable for Torus {}
+
+struct Capsule {
+  height: Vec3,
+  radius: f32,
+}
+impl Meshable for Capsule {}
+
+// A 3d frustum used to represent the volume rendered by a camera, defined by the 6 planes that set the frustum limits.
+struct Frustum{
+  near: Plane,
+  far: Plane,
+  top: Plane,
+  bottom: Plane,
+  left: Plane,
+  right: Plane,
+}
+impl Meshable for Frustum {}
+
+```
+
+### 3D Bounding Volumes
+
+```rust
+
+/// A spherical bounding volume fully defined in space
+struct SphereBV {
+  sphere: Sphere,
+  origin: Point,
+}
+impl BoundingVolume for SphereBV ()
+
+struct CylinderBV {
+  cylinder: Cylinder,
+  origin: Point,
+  rotation: Quat,
+}
+impl BoundingVolume for CylinderBV {}
+
+struct CapsuleBV {
+  capsule: Capsule,
+  origin: Point,
+}
+impl BoundingVolume for CapsuleBV {}
+
+// 3D Axis Aligned Bounding Box defined by its extents, useful for fast intersection checks and frustum culling.
+struct Aabb {
+  min: Vec3
+  max: Vec3
+}
+impl Meshable for Aabb {}
+
+// 3D Oriented Bounding Box
+struct Obb {
+  origin: Point,
+  orthonormal_basis: Mat3,
+  half_extents: Vec3,
+}
+impl Meshable for Obb {}
+```
+
+### 2D Geometry Types
+
+These types only exist in 2d space: their dimensions and location are only defined in `x` and `y` unlike their 3d counterparts. These types are suffixed with "2d" to disambiguate from the 3d types in user code, guide users to using 3d types by default, and remove the need for name-spacing the 2d and 3d types when used in the same scope.
+
+```rust
+/// Point in 2D space
+struct Point2d(Vec2)
+
+/// Vector direction in 2D space that is guaranteed to be normalized through its getter/setter.
+struct Direction2d(Vec2)
+
+
+/// Unbounded line in 2D space with direction
+struct Line2d { 
+  point: Point2d, 
+  normal: Direction2d,
+}
+
+/// A newtype that differentiates a line from a ray, where a line is an infinite axis and a ray is directional half-line.
+struct Ray2d(Line2d);
+
+/// Line in 2D space bounded by two points
+struct LineSegment2d { 
+  start: Point2d, 
+  end: Point2d,
+}
+
+/// A line list represented by an ordered list of vertices in 2d space
+struct LineList2d<const N: usize>{
```

Comment:
Does this have any different behavior than if it were called PolyLine2d with an option to be closed? As in are there any significant reasons to call it a LineList?